### ISSUE TITLE
[Llama 70B Galaxy] 32 users in batch with same inputs produce different outputs

### ISSUE THREAD

**Issue by rdraskicTT** [tenstorrent/tt-metal#35017]
Ops inside model + sampling execution advance PRNG state so batch_size=32 for same inputs produce different outputs due to different PRNG state across the cores.

**Comment by djordje-tt** [comment]
- There are many ops that can advance the PRNG state on each core. Since these ops may execute on a variable number of cores, it’s impossible to keep PRNGs synchronized across cores, even if they are reset with the same seed and there are ops in between reset and rand_tile call. I discussed this with [@Nikola Velickovic](https://tenstorrent.enterprise.slack.com/team/U08HH1TFMSB) and [@Anil Mahmud](https://tenstorrent.enterprise.slack.com/team/U08JC875AVA).

As a result, we cannot reliably use PRNGs when their state can be modified by other ops. This is hard to track because it’s not a single op, and it’s also data-dependent when the PRNG state advances.

Given this, there are two possible approaches:

- Hacky workaround (short-term): Reset the seed before every rand_tile, but not with a fixed seed. After calling reset_seed(value) and rand_tile, extract one element from the generated tile and use it as the seed for the next iteration. This guarantees the same sequence of random tiles regardless of execution order or user, and ensures no ops interfere between rand_tile_init and rand_tile. This would require modifying the current implementation by adding ttnn.manual_seed inside the trace just before sampling op and passing seed values via a preallocated tensor.

- Proper solution (long-term): Implement a software-based RNG as proposed by [@Chris Maryan](https://tenstorrent.enterprise.slack.com/team/U08411R54Q2). This will likely take more time and will be discussed separately with him.

**Comment by djordje-tt** [comment]
Added workaround with keeping seeds on host. Tested with vllm and tt-metal with decode sampling only and it's producing same outputs for all users. Currently verifying with different sampling params that it's passing the tests. 
Next steps: 
- apply workaround for Llama70b-glx with prefill+decode sampling
- apply workaround for Llama3.1-8b
Majority of work is to test and add simple modifications since sampling module is reused and most of implementation is shared there. Just couple of copy-pastes across generators + testing across tt-metal, vllm and inference-server is major effort now.

**Comment by djordje-tt** [comment]
[Skip to summary]
After inspection of prefill+decode seems like latest tests introduced bugs with non-batched prefill (e.g. 2 users) logits before sampling in prefill are different for same inputs across users. Also batched_prefill for 16 users is failing. 

Apart from these 2 issues we are seeing correct outputs for batched prefill while for non-batched prefill we are not seeing correct outputs which might be caused by the same issue as above mentioned. This is still needed to be investigated. I will try to do that tonight and tomorrow morning but cannot guarantee we will make it until the end of the year. 

Summary of bugs: 
- batched_prefill for 16 users is failing in syntax (might be introduced with my commit)
- non-batched prefill logits before sampling are different (might be issues with persistent buffer since we are running lm_head for multiple users consecutively without running model itself. 
- previous issue probably causes later users to diverge. Once non-batched prefill logits are fixed this could be tested.

**Comment by djordje-tt** [comment]
Latest summary: 
- Corrupted prefill logits with trace enabled - @kpaigwar  investigating
- Update sampling params to be correct on empty slots - small fix - @rdraskicTT  to do
- Difference in output responses when using batched vs non-batched prefill - @rdraskicTT  investigating
- Apply seeding impl from Llama glx -> TTT, I didn’t have time to finish this due to other bugs I was fixing in the meantime. @sraizada-tt  to take a look on Monday.

@uaydonat, if you think someone else should jump in or take over feel free to update issue assignees.

**Comment by kpaigwar** [comment]
> Corrupted prefill logits with trace enabled - @kpaigwar investigating

In non-batched prefill mode, previous user prefill logits' were getting corrupted by next user prefill trace execution. This could be happening as we are allocating new tensor through clone right after prefill trace execution. The captured trace might be using the same buffer address space of cloned tensor and corrupting it.

Solution: Create a list of persistent tensor for accumulation before trace capture and use ttnn.copy to update corresponding persistent output tensor.

@jonathansuTT and I have a fix pushed here https://github.com/tenstorrent/tt-metal/commit/aecdc31bc05b1c680d9b078e4b679fdaa11b45af

**Comment by kpaigwar** [comment]
Took a look at the issue

> Difference in output responses when using batched vs non-batched prefill - @rdraskicTT investigating

I also see different response for same prompt and seed value. But they are not garbage
Looks like first token generated itself is different but its not sampling issue. tt-logits after LM head itself shows that they are different.

Here, I fetched device 0 tensor chuck for both prefills and sliced to remove any padding

DEBUG [BATCHED]: Saved first user logits, shape=torch.Size([16032])
DEBUG [BATCHED]: logits stats: min=-9.2500, max=14.1250, mean=0.0826

DEBUG [NON-BATCHED]: Saved first user logits, shape=torch.Size([16032])
DEBUG [NON-BATCHED]: logits stats: min=-9.2500, max=13.8750, mean=0.0730
DEBUG [COMPARE]: abs diff: min=0.000000, max=0.687500, mean=0.123671

@sraizada-tt @rdraskicTT is this expected behavior as perhaps both implementation could be using some different kernels

**Comment by ekuslysTT** [comment]
Editing labels as this is tracked through the parent issue
